django/conf/__init__.py

```diff
@@ -129,6 +137,20 @@ def configured(self):
         """Return True if the settings have already been configured."""
         return self._wrapped is not empty
 
+    @property
+    def USE_L10N(self):
+        stack = traceback.extract_stack()
+        # Show a warning if the setting is used outside of Django.
+        # Stack index: -1 this line, -2 the caller.
+        filename, _, _, _ = stack[-2]
+        if not filename.startswith(os.path.dirname(django.__file__)):
+            warnings.warn(
+                USE_L10N_DEPRECATED_MSG,
+                RemovedInDjango50Warning,
+                stacklevel=2,
+            )
+        return self.__getattr__('USE_L10N')
```

Comment:
Any idea what the "cost" of this is? 

ie: because all template output runs through `render_value_in_context` -> `localize` which then dispatches to any of `number_format` / `date_format` / `time_format` each of which _may_ call `settings.USE_L10N` depending on if the `context.use_l10n` value (and I confess I can't remember when/where the details of _that_), each of those values (decimals/ints/floats/datetime.*) is unavoidably going to be slower (if/when `use_l10n` is `None`), and I wonder by how much? And is it avoidable? (eg: `cached_property` or what-have-you)

Reply:
I added an internal hook `_USE_L10N_INTERNAL` that should be enough to avoid checking a traceback on hot paths.